pkg/util/validation/validate.go

```diff
@@ -238,3 +261,13 @@ func formatLabelSet(ls []client.LabelAdapter) string {
 
 	return fmt.Sprintf("%s{%s}", metricName, strings.Join(labelStrings, ", "))
 }
+
+func DeletePerUserValidationMetrics(userID string) {
```

Comment:
Unfortunately this is a bit fragile, but I don't have smart ideas. The problem is that we rely on `allDiscardedSampleReasons` and `allDiscardedMetadataReasons` but they will de-sync as soon as we'll add another reason (we'll forget to add it to the list of reasons, especially if done in another package).

What if we discuss with Beorn the option to add a delete function which takes in input only some labels and delete all metrics matching it?

Reply:
> What if we discuss with Beorn the option to add a delete function which takes in input only some labels and delete all metrics matching it?

This would also help us to clean `cortex_distributor_deduped_samples_total` in distributor, which we cannot currently remove as we don't keep track of values for `cluster` label.

Reply:
Thanks for bringing this to my attention. I don't remember anyone has brought this up before. Which is in hindsight a bit surprising, given that there had been quite a few discussions about partial label specification (which lead to currying). I guess deletion just doesn't happen that often.

I collected my thoughts in https://github.com/prometheus/client_golang/issues/834 (which is at the same time the formal feature request for what you want).

Reply:
Given we're not expected to delete series too frequently, using `Collect()` to find out metric labels may be a viable option for us. WDYT @pstibrany ?

Reply:
Yes, for this case, I think it's a good workaround. If there is different solution in the future, we can use that.

Reply:
Implemented collect-based approach to find labels and deleting all user-specific labels.